diff -ru pjsip/include/pjsip/sip_uri.h pjsip/include/pjsip/sip_uri.h
--- pjsip/include/pjsip/sip_uri.h	2013-06-19 06:47:43.000000000 +0000
+++ pjsip/include/pjsip/sip_uri.h	2018-11-21 15:38:37.687213747 +0000
@@ -294,7 +294,7 @@
  * @param uri	    The URI to print.
  * @param buf	    The buffer.
  * @param size	    Size of the buffer.
- * @return	    Length printed.
+ * @return          Length printed if successful, negative value if failed.
  */
 PJ_INLINE(int) pjsip_uri_print(pjsip_uri_context_e context,
 			       const void *uri,
diff -ru pjsip/src/pjsua-lib/pjsua_call.c pjsip/src/pjsua-lib/pjsua_call.c
--- pjsip/src/pjsua-lib/pjsua_call.c	2017-01-19 10:31:38.000000000 +0000
+++ pjsip/src/pjsua-lib/pjsua_call.c	2018-11-21 15:39:15.960282508 +0000
@@ -1820,6 +1820,8 @@
 					       dlg->local.contact->uri,
 					       info->local_contact.ptr,
 					       sizeof(info->buf_.local_contact));
+	if (info->local_contact.slen < 0)
+        info->local_contact.slen = 0;
 
     /* remote info */
     info->remote_info.ptr = info->buf_.remote_info;
diff -ru pjsip/src/pjsua2/util.hpp pjsip/src/pjsua2/util.hpp
--- pjsip/src/pjsua2/util.hpp	2014-01-16 05:30:46.000000000 +0000
+++ pjsip/src/pjsua2/util.hpp	2018-11-21 15:36:42.066002629 +0000
@@ -36,7 +36,7 @@
 
 inline string pj2Str(const pj_str_t &input_str)
 {
-    if (input_str.ptr)
+    if (input_str.ptr && input_str.slen>0)
 	return string(input_str.ptr, input_str.slen);
     return string();
 }
