diff -Naur pjproject-2.4.5/aconfigure.ac pjproject-2.4.5/aconfigure.ac
--- pjproject-2.4.5/aconfigure.ac	2015-12-27 11:12:49.000000000 +0000
+++ pjproject-2.4.5/aconfigure.ac	2015-12-27 11:18:36.000000000 +0000
@@ -1255,6 +1255,20 @@
 			       )
 	      ])

+dnl ########################################################
+dnl # G.729 support
+dnl #
+AC_SUBST(ac_has_g729_codec)
+AC_ARG_ENABLE(g729_codec,
+    AC_HELP_STRING([--enable-g729-codec],
+       [Enable g729 support (derived from ITU implementation).]),
+        [ac_has_g729_codec=1]
+        AC_DEFINE(PJMEDIA_HAS_G729_CODEC,1)
+        AC_MSG_RESULT([Checking if g729 codec is disabled...no]),
+    [ac_has_g729_codec=]
+    AC_DEFINE(PJMEDIA_HAS_G729_CODEC,0)
+    AC_MSG_RESULT([Checking if g729 codec is disabled...yes])
+    )

 dnl ########################################################
 dnl # Intel IPP support
diff -Naur pjproject-2.4.5/build.mak.in pjproject-2.4.5/build.mak.in
--- pjproject-2.4.5/build.mak.in	2015-12-27 11:12:49.000000000 +0000
+++ pjproject-2.4.5/build.mak.in	2015-12-27 11:14:10.000000000 +0000
@@ -106,6 +106,11 @@
 endif
 endif

+ifeq (@ac_has_g729_codec@,1)
+APP_THIRD_PARTY_LIBS += -lg729codec-$(TARGET_NAME)
+APP_THIRD_PARTY_LIB_FILES += $(PJ_DIR)/third_party/lib/libg729codec-$(LIB_SUFFIX)
+endif
+
 ifneq ($(findstring pa,@ac_pjmedia_snd@),)
 ifeq (@ac_external_pa@,1)
 # External PA
diff -Naur pjproject-2.4.5/pjmedia/build/os-auto.mak.in pjproject-2.4.5/pjmedia/build/os-auto.mak.in
--- pjproject-2.4.5/pjmedia/build/os-auto.mak.in	2015-12-27 11:12:50.000000000 +0000
+++ pjproject-2.4.5/pjmedia/build/os-auto.mak.in	2015-12-27 11:15:24.000000000 +0000
@@ -72,6 +72,7 @@
 AC_NO_ILBC_CODEC=@ac_no_ilbc_codec@
 AC_NO_G722_CODEC=@ac_no_g722_codec@
 AC_NO_G7221_CODEC=@ac_no_g7221_codec@
+AC_HAS_G729_CODEC=@ac_has_g729_codec@
 AC_NO_OPENCORE_AMRNB=@ac_no_opencore_amrnb@
 AC_NO_OPENCORE_AMRWB=@ac_no_opencore_amrwb@

@@ -128,6 +129,13 @@
 export G7221_CFLAGS += -I$(THIRD_PARTY)
 endif

+ifeq ($(AC_HAS_G729_CODEC),1)
+export CFLAGS += -I$(THIRD_PARTY)/g729/include -DPJMEDIA_HAS_G729_CODEC=1
+export CODEC_OBJS += g729.o
+else
+export CFLAGS += -DPJMEDIA_HAS_G729_CODEC=0
+endif
+
 ifeq ($(AC_NO_OPENCORE_AMRNB),1)
 export CFLAGS += -DPJMEDIA_HAS_OPENCORE_AMRNB_CODEC=0
 else
diff -Naur pjproject-2.4.5/pjmedia/include/pjmedia-codec/config_auto.h.in pjproject-2.4.5/pjmedia/include/pjmedia-codec/config_auto.h.in
--- pjproject-2.4.5/pjmedia/include/pjmedia-codec/config_auto.h.in	2015-12-27 11:12:51.000000000 +0000
+++ pjproject-2.4.5/pjmedia/include/pjmedia-codec/config_auto.h.in	2015-12-27 11:17:16.000000000 +0000
@@ -69,6 +69,11 @@
 #undef PJMEDIA_HAS_G7221_CODEC
 #endif

+/* G729 codec */
+#ifndef PJMEDIA_HAS_G729_CODEC
+#undef PJMEDIA_HAS_G729_CODEC
+#endif
+
 /* OpenCORE AMR-NB codec */
 #ifndef PJMEDIA_HAS_OPENCORE_AMRNB_CODEC
 #undef PJMEDIA_HAS_OPENCORE_AMRNB_CODEC
diff -Naur pjproject-2.4.5/pjmedia/include/pjmedia-codec/config.h pjproject-2.4.5/pjmedia/include/pjmedia-codec/config.h
--- pjproject-2.4.5/pjmedia/include/pjmedia-codec/config.h	2015-12-27 11:12:51.000000000 +0000
+++ pjproject-2.4.5/pjmedia/include/pjmedia-codec/config.h	2015-12-27 11:16:28.000000000 +0000
@@ -96,6 +96,12 @@
 #   define PJMEDIA_HAS_G722_CODEC    1
 #endif

+/**
+* Unless specified otherwise, G.729 codec is not included by default.
+*/
+#ifndef PJMEDIA_HAS_G729_CODEC
+#   define PJMEDIA_HAS_G729_CODEC    0
+#endif

 /**
  * Default G.722 codec encoder and decoder level adjustment. The G.722
diff -Naur pjproject-2.4.5/pjmedia/include/pjmedia-codec.h pjproject-2.4.5/pjmedia/include/pjmedia-codec.h
--- pjproject-2.4.5/pjmedia/include/pjmedia-codec.h	2015-12-27 11:12:51.000000000 +0000
+++ pjproject-2.4.5/pjmedia/include/pjmedia-codec.h	2015-12-27 11:17:48.000000000 +0000
@@ -33,6 +33,7 @@
 #include <pjmedia-codec/ilbc.h>
 #include <pjmedia-codec/g722.h>
 #include <pjmedia-codec/g7221.h>
+#include <pjmedia-codec/g729.h>
 #include <pjmedia-codec/ipp_codecs.h>
 #include <pjmedia-codec/opencore_amr.h>
 #include <pjmedia-codec/openh264.h>
diff -Naur pjproject-2.4.5/pjmedia/src/pjmedia-codec/audio_codecs.c pjproject-2.4.5/pjmedia/src/pjmedia-codec/audio_codecs.c
--- pjproject-2.4.5/pjmedia/src/pjmedia-codec/audio_codecs.c	2015-12-27 11:12:51.000000000 +0000
+++ pjproject-2.4.5/pjmedia/src/pjmedia-codec/audio_codecs.c	2015-12-27 11:20:05.000000000 +0000
@@ -100,6 +100,13 @@
 	return status;
 #endif /* PJMEDIA_HAS_G7221_CODEC */

+#if PJMEDIA_HAS_G729_CODEC
+    status = pjmedia_codec_g729_init(endpt);
+    if (status != PJ_SUCCESS) {
+        return status;
+    }
+#endif /* PJMEDIA_HAS_G729_CODEC */
+
 #if PJMEDIA_HAS_L16_CODEC
     /* Register L16 family codecs */
     status = pjmedia_codec_l16_init(endpt, 0);
diff -Naur pjproject-2.4.5/third_party/build/os-auto.mak.in pjproject-2.4.5/third_party/build/os-auto.mak.in
--- pjproject-2.4.5/third_party/build/os-auto.mak.in	2015-12-27 11:12:54.000000000 +0000
+++ pjproject-2.4.5/third_party/build/os-auto.mak.in	2015-12-27 11:20:46.000000000 +0000
@@ -23,6 +23,10 @@
 DIRS += g7221
 endif

+ifeq (@ac_has_g729_codec@,1)
+DIRS += g729
+endif
+
 ifneq ($(findstring pa,@ac_pjmedia_snd@),)
 ifeq (@ac_external_pa@,1)
 # External PA
