diff -Naur ./pjproject-2.12.1/pjsip/src/pjsua-lib/pjsua_acc.c ./pjproject-2.12.1-patched/pjsip/src/pjsua-lib/pjsua_acc.c
--- ./pjsip/src/pjsua-lib/pjsua_acc.c	2023-11-24 15:02:28
+++ ./pjsip/src/pjsua-lib/pjsua_acc.c	2023-11-27 14:41:25
@@ -1580,7 +1580,7 @@
 	      acc->cfg.reg_contact_params.slen +
 	      acc->cfg.reg_contact_uri_params.slen +
 	      (need_outbound?
-	       (acc->rfc5626_instprm.slen + acc->rfc5626_regprm.slen): 0);
+	       (acc->rfc5626_instprm.slen + acc->rfc5626_regprm.slen): 1);
 	if (len > acc->contact.slen) {
 	    reg_contact.ptr = (char*) pj_pool_alloc(acc->pool, len);
 
